README.md
# Sauvignon-in-Stardew
A Stardew Valley Mod


RELEASE-NOTES.md
[← back to readme](README.md)

# Release notes
## Upcoming release
* Winery data is now stored in the save file. (Existing wineries will be migrated next time you save.
* Updated for Stardew Valley 1.3.32, SMAPI 2.9, and the upcoming SMAPI 3.0.
* Fixed errors due to broken syntax in the i18n files.
* Fixed errors when a winery is outside the map bounds (e.g. because a custom map was removed).
* Fixed lag inside the winery for some players.

## 1.9
* HOPEFULLY fixed Distiller craft bug.

## 1.8
* Added compatibility with SpaceCore.
* Fixed compatibility with All Professions.

## 1.7
* Added compatibility with mods like All Professions.
* Added mouse pointer to SkillsPage hovering.
* Fixed compatibility issue with CJB Item Spawner.

## 1.6
* Fixed the Distiller profession.
* Removed test code that affected kegs.

## 1.5
* Added music for the Winery. :D
* Added option to disable Distiller profession.
* Fixed Distiller profession not giving bonuses.
* Fixed not being able to select Distiller profession.
* Fixed Distiller profession taking away money.

## 1.4
* Added Distiller profession.
* Added Distiller Craft category.
* Changed hover tooltip to show item name as well as time remaining.
* Reduced clutter in the winery interior.
* Fixed small bugs.

## 1.3
* Added MTN compatibility.
* Fixed some small bugs.
* Optimised code.

## 1.2
* Fixed kegs not getting the 30% bonus while sleeping.
* Fixed seasonal textures not loading.
* Added update keys.

## 1.1
* Added seasonal building textures. Art provided by a person in the credits. :D
* Fixed an error with moving the Winery building in Robin's Carpenter Shop.
* Preparing for upcoming Distilled Craft category (beer, mead, IPA, and wines).

## 1.0
* Initial release.


testing2.cs
//Global Variables
public Texture2D Winery_outdoors;
public Map kegRoom_indoors;

//Inside Entry Method
Winery_outdoors = helper.Content.Load<Texture2D>($"assets/Winery_outside_{Game1.currentSeason}.png", ContentSource.ModFolder);
kegRoom_indoors = helper.Content.Load<Map>("assets/Winery2.tbin", ContentSource.ModFolder);
MenuEvents.MenuChanged += MenuEvents_MenuChanged;

//Inside Class
public static bool IsMagical(IClickableMenu carpenterMenu)
{
    return helper.Reflection.GetField<bool>(carpenterMenu, "magicalConstruction").GetValue();
}

public static bool HasBluePrint(IClickableMenu carpenterMenu, string blueprintName)
{
    return GetBluePrints(carpenterMenu).Exists(bluePrint => bluePrint.name == blueprintName);
}

public static void SetBluePrintField(BluePrint bluePrint, string field, object value)
{
    helper.Reflection.GetField<object>(bluePrint, field).SetValue(value);
}

if (e.NewMenu is CarpenterMenu)
{
  if (!Game1.getFarm().buildings.Any(building => building.buildingType.Value == "Winery") && !IsMagical(e.NewMenu) && !HasBluePrint(e.NewMenu, "Winery2"))
    {
        BluePrint kegRoomBluePrint = new BluePrint("Slime Hutch")
        {
            name = "Winery2",
            displayName = "Keg Room",
            description = "Adds a room to your Winery that houses huge kegs able to process large quantities of products.",
            daysToConstruct = 2,
            moneyRequired = 450000,
            blueprintType = "Upgrades",
            nameOfBuildingToUpgrade = "Winery"
        };
        kegRoomBluePrint.itemsRequired.Clear();
        kegRoomBluePrint.itemsRequired.Add(709, 200);//200
        kegRoomBluePrint.itemsRequired.Add(335, 150);//150
        kegRoomBluePrint.itemsRequired.Add(388, 500);//500

        SetBluePrintField(kegRoomBluePrint, "textureName", "Buildings\\Winery2");
        SetBluePrintField(kegRoomBluePrint, "texture", Game1.content.Load<Texture2D>(kegRoomBluePrint.textureName));
    }
}

//Loader
public bool CanLoad<T>(IAssetInfo asset)
{
    if (asset.AssetNameEquals("Buildings\\Winery2"))
    {
        return true;
    }
    else if (asset.AssetNameEquals("Maps/Winery2"))
    {
        return true;
    }
    return false;
}

public T Load<T>(IAssetInfo asset)
{
    if (asset.AssetNameEquals("Buildings\\Winery2"))
    {
        return (T)(object)Winery_outdoors;
    }
    else if (asset.AssetNameEquals("Maps/Winery2"))
    {
        return (T)(object)kegRoom_indoors;
    }
    return (T)(object)null;
}


SauvignonInStardew/DistillerMenu.cs
using System.Collections.Generic;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using StardewModdingAPI;
using StardewValley;
using StardewValley.Menus;

namespace SauvignonInStardew
{
    internal class DistillerMenu : LevelUpMenu
    {
        /*********
        ** Fields
        *********/
        /// <summary>Provides an API for checking and changing input state.</summary>
        private readonly IInputHelper Input;

        private Color FirstProfessionColor;
        private Color SecondProfessionColor;
        private Color ThirdProfessionColor;

        private readonly List<string> FirstProfessionDescription = new List<string>() { "Artisan", "Artisan goods (cheese, truffle oil, cloth, etc.) worth 40% more." };

        private readonly List<string> SecondProfessionDescription = new List<string>() { "Agriculturist", "All crops grow 10% faster." };

        private readonly List<string> ThirdProfessionDescription = new List<string>() { "Distiller", "Alcohol (beer, wine, etc.) worth 40% more." };

        private readonly string Title = "Level 10 Farming";

        private readonly string Text = "Choose a profession:";

        private readonly List<int> ProfessionsToChoose = new List<int> { 4, 5, 77 };

        private readonly Rectangle SourceRectForLevelIcon = new Rectangle(0, 0, 16, 16);

        private readonly List<TemporaryAnimatedSprite> LittleStars = new List<TemporaryAnimatedSprite>();

        private int TimerBeforeStart;

        private readonly Texture2D Icon;


        /*********
        ** Public methods
        *********/
        public DistillerMenu(Texture2D icon, IInputHelper input)
        {
            this.Icon = icon;
            this.Input = input;
            this.TimerBeforeStart = 250;
            this.isActive = true;
            this.height = 512;
            this.width = 1344;

            Game1.player.completelyStopAnimatingOrDoingAction();
            this.xPositionOnScreen = 100;
            this.yPositionOnScreen = (int)(Game1.viewport.Height * (1.0 / Game1.options.zoomLevel)) / 4;

            //this.populateClickableComponentList();
            this.allClickableComponents = new List<ClickableComponent>
            {
                // first profession
                new ClickableComponent(new Rectangle(this.xPositionOnScreen, this.yPositionOnScreen + 128, this.width / 3, this.height), "")
                {
                    myID = 102,
                    rightNeighborID = 103
                },

                // second profession
                new ClickableComponent(new Rectangle(this.width / 3 + this.xPositionOnScreen, this.yPositionOnScreen + 128, this.width / 3, this.height), "")
                {
                    myID = 103,
                    leftNeighborID = 102
                },

                // third profession
                new ClickableComponent(new Rectangle((this.width / 3) * 2 + this.xPositionOnScreen, this.yPositionOnScreen + 128, this.width / 3, this.height), "")
                {
                    myID = 104,
                    leftNeighborID = 103
                }
            };
            this.snapToDefaultClickableComponent();
        }

        public override void update(GameTime time)
        {
            if (!this.isActive)
            {
                this.exitThisMenu();
            }

            //stars
            for (int index = this.LittleStars.Count - 1; index >= 0; --index)
            {
                if (this.LittleStars[index].update(time))
                    this.LittleStars.RemoveAt(index);
            }
            if (Game1.random.NextDouble() < 0.03)
            {
                Vector2 position = new Vector2(0.0f, Game1.random.Next(this.yPositionOnScreen - 128, this.yPositionOnScreen - 4) / 20 * 4 * 5 + 32);
                position.X = Game1.random.NextDouble() >= 0.5 ? Game1.random.Next(this.xPositionOnScreen + this.width / 2 + 116, this.xPositionOnScreen + this.width - 160) : Game1.random.Next(this.xPositionOnScreen + this.width / 2 - 228, this.xPositionOnScreen + this.width / 2 - 132);
                if (position.Y < (double)(this.yPositionOnScreen - 64 - 8))
                    position.X = Game1.random.Next(this.xPositionOnScreen + this.width / 2 - 116, this.xPositionOnScreen + this.width / 2 + 116);
                position.X = (float)(position.X / 20.0 * 4.0 * 5.0);
                this.LittleStars.Add(new TemporaryAnimatedSprite("LooseSprites\\Cursors", new Rectangle(364, 79, 5, 5), 80f, 7, 1, position, false, false, 1f, 0.0f, Color.White, 4f, 0.0f, 0.0f, 0.0f)
                {
                    local = true
                });
            }
            //end stars

            if (this.TimerBeforeStart > 0)
            {
                this.TimerBeforeStart -= time.ElapsedGameTime.Milliseconds;
            }
            else
            {
                this.FirstProfessionColor = Game1.textColor;
                this.SecondProfessionColor = Game1.textColor;
                this.ThirdProfessionColor = Game1.textColor;
                if (Game1.getMouseY() > this.yPositionOnScreen + 192 && Game1.getMouseY() < this.yPositionOnScreen + this.height)
                {
                    if (Game1.getMouseX() > this.xPositionOnScreen && Game1.getMouseX() < this.xPositionOnScreen + this.width / 3)
                    {
                        this.FirstProfessionColor = Color.Green;
                        //if ((Game1.input.GetMouseState().LeftButton == ButtonState.Pressed && this.oldMouseState.LeftButton == ButtonState.Released || Game1.options.gamepadControls && (Game1.input.GetGamePadState().IsButtonDown(Buttons.A) && !Game1.oldPadState.IsButtonDown(Buttons.A))) && this.readyToClose())
                        if ((this.Input.IsDown(SButton.MouseLeft)) || (Game1.options.gamepadControls && this.Input.IsDown(SButton.A)))
                        {
                            if (!Game1.player.professions.Contains(4))
                            {
                                Game1.player.professions.Add(this.ProfessionsToChoose[0]);
                                this.getImmediateProfessionPerk(this.ProfessionsToChoose[0]);
                                this.isActive = false;
                                this.informationUp = false;
                                this.isProfessionChooser = false;
                            }
                        }
                    }
                    else if (Game1.getMouseX() > this.xPositionOnScreen + this.width / 3 && Game1.getMouseX() < this.xPositionOnScreen + this.width / 3 + this.width / 3)
                    {
                        this.SecondProfessionColor = Color.Green;
                        if ((this.Input.IsDown(SButton.MouseLeft)) || (Game1.options.gamepadControls && this.Input.IsDown(SButton.A)))
                        {
                            if (!Game1.player.professions.Contains(5))
                            {
                                Game1.player.professions.Add(this.ProfessionsToChoose[1]);
                                this.getImmediateProfessionPerk(this.ProfessionsToChoose[1]);
                                this.isActive = false;
                                this.informationUp = false;
                                this.isProfessionChooser = false;
                            }
                        }
                    }
                    else if (Game1.getMouseX() > this.xPositionOnScreen + this.width / 3 + this.width / 3 && Game1.getMouseX() < this.xPositionOnScreen + this.width / 3 + this.width / 3 + this.width / 3)
                    {
                        this.ThirdProfessionColor = Color.Green;
                        if ((this.Input.IsDown(SButton.MouseLeft)) || (Game1.options.gamepadControls && this.Input.IsDown(SButton.A)))
                        {
                            if (!Game1.player.professions.Contains(77))
                            {
                                Game1.player.professions.Add(this.ProfessionsToChoose[2]);
                                this.getImmediateProfessionPerk(this.ProfessionsToChoose[2]);
                                this.isActive = false;
                                this.informationUp = false;
                                this.isProfessionChooser = false;
                            }
                        }
                    }
                }
            }
        }

        public override void draw(SpriteBatch b)
        {
            //stars
            b.Draw(Game1.fadeToBlackRect, new Rectangle(0, 0, Game1.viewport.Width, Game1.viewport.Height), Color.Black * 0.5f);
            foreach (TemporaryAnimatedSprite littleStar in this.LittleStars)
                littleStar.draw(b);
            b.Draw(Game1.mouseCursors, new Vector2(this.xPositionOnScreen + this.width / 2 - 116, this.yPositionOnScreen - 32 + 12), new Rectangle(363, 87, 58, 22), Color.White, 0.0f, Vector2.Zero, 4f, SpriteEffects.None, 1f);

            //top info and actual menu
            Game1.drawDialogueBox(this.xPositionOnScreen, this.yPositionOnScreen, this.width, this.height, false, true);

            this.drawHorizontalPartition(b, this.yPositionOnScreen + 192);

            this.drawVerticalIntersectingPartition(b, this.xPositionOnScreen + this.width / 3 - 32, this.yPositionOnScreen + 192);
            this.drawVerticalIntersectingPartition(b, this.xPositionOnScreen + this.width / 3 - 32 + this.width / 3, this.yPositionOnScreen + 192);

            Utility.drawWithShadow(b, Game1.buffsIcons, new Vector2(this.xPositionOnScreen + IClickableMenu.spaceToClearSideBorder + IClickableMenu.borderWidth, this.yPositionOnScreen + IClickableMenu.spaceToClearTopBorder + 16), this.SourceRectForLevelIcon, Color.White, 0.0f, Vector2.Zero, 4f, false, 0.88f);

            b.DrawString(Game1.dialogueFont, this.Title, new Vector2(this.xPositionOnScreen + this.width / 2 - Game1.dialogueFont.MeasureString(this.Title).X / 2f, this.yPositionOnScreen + IClickableMenu.spaceToClearTopBorder + 16), Game1.textColor);

            Utility.drawWithShadow(b, Game1.buffsIcons, new Vector2(this.xPositionOnScreen + this.width - IClickableMenu.spaceToClearSideBorder - IClickableMenu.borderWidth - 64, this.yPositionOnScreen + IClickableMenu.spaceToClearTopBorder + 16), this.SourceRectForLevelIcon, Color.White, 0.0f, Vector2.Zero, 4f, false, 0.88f);

            b.DrawString(Game1.smallFont, this.Text, new Vector2(this.xPositionOnScreen + this.width / 2 - Game1.smallFont.MeasureString(this.Text).X / 2f, this.yPositionOnScreen + 64 + IClickableMenu.spaceToClearTopBorder), Game1.textColor);

            //first profession
            //title
            b.DrawString(Game1.dialogueFont, this.FirstProfessionDescription[0], new Vector2(this.xPositionOnScreen + IClickableMenu.spaceToClearSideBorder + 32, this.yPositionOnScreen + IClickableMenu.spaceToClearTopBorder + 160), this.FirstProfessionColor);

            //icon
            b.Draw(Game1.mouseCursors, new Vector2(this.xPositionOnScreen + IClickableMenu.spaceToClearSideBorder + this.width / 3 - 112, this.yPositionOnScreen + IClickableMenu.spaceToClearTopBorder + 160 - 16), new Rectangle(this.ProfessionsToChoose[0] % 6 * 16, 624 + this.ProfessionsToChoose[0] / 6 * 16, 16, 16), Color.White, 0.0f, Vector2.Zero, 4f, SpriteEffects.None, 1f);

            //description
            for (int index = 1; index < this.FirstProfessionDescription.Count; ++index)
            {
                b.DrawString(Game1.smallFont, Game1.parseText(this.FirstProfessionDescription[index], Game1.smallFont, this.width / 3 - 64), new Vector2(this.xPositionOnScreen - 4 + IClickableMenu.spaceToClearSideBorder + 32, this.yPositionOnScreen + IClickableMenu.spaceToClearTopBorder + 128 + 8 + 64 * (index + 1)), this.FirstProfessionColor);
            }

            //second profession
            //title
            b.DrawString(Game1.dialogueFont, this.SecondProfessionDescription[0], new Vector2(this.xPositionOnScreen + IClickableMenu.spaceToClearSideBorder + this.width / 3, this.yPositionOnScreen + IClickableMenu.spaceToClearTopBorder + 160), this.SecondProfessionColor);

            //icon
            b.Draw(Game1.mouseCursors, new Vector2(this.xPositionOnScreen + IClickableMenu.spaceToClearSideBorder + (this.width / 3) * 2 - 128, this.yPositionOnScreen + IClickableMenu.spaceToClearTopBorder + 160 - 16), new Rectangle(this.ProfessionsToChoose[1] % 6 * 16, 624 + this.ProfessionsToChoose[1] / 6 * 16, 16, 16), Color.White, 0.0f, Vector2.Zero, 4f, SpriteEffects.None, 1f);

            //description
            for (int index = 1; index < this.SecondProfessionDescription.Count; ++index)
            {
                b.DrawString(Game1.smallFont, Game1.parseText(this.SecondProfessionDescription[index], Game1.smallFont, this.width / 3 - 48), new Vector2(this.xPositionOnScreen - 4 + IClickableMenu.spaceToClearSideBorder + this.width / 3, this.yPositionOnScreen + IClickableMenu.spaceToClearTopBorder + 128 + 8 + 64 * (index + 1)), this.SecondProfessionColor);
            }

            //third profession
            //title
            b.DrawString(Game1.dialogueFont, this.ThirdProfessionDescription[0], new Vector2(this.xPositionOnScreen + IClickableMenu.spaceToClearSideBorder + this.width / 3 + this.width / 3, this.yPositionOnScreen + IClickableMenu.spaceToClearTopBorder + 160), this.ThirdProfessionColor);

            //icon
            b.Draw(this.Icon, new Vector2(this.xPositionOnScreen + IClickableMenu.spaceToClearSideBorder + (this.width / 3) * 3 - 128, this.yPositionOnScreen + IClickableMenu.spaceToClearTopBorder + 160 - 16), new Rectangle(0, 0, 16, 16), Color.White, 0.0f, Vector2.Zero, 4f, SpriteEffects.None, 1f);//this.professionsToChoose[2] % 6 * 16, 624 + this.professionsToChoose[2] / 6 * 16

            //description
            for (int index = 1; index < this.ThirdProfessionDescription.Count; ++index)
            {
                b.DrawString(Game1.smallFont, Game1.parseText(this.ThirdProfessionDescription[index], Game1.smallFont, (this.width / 3 - 48)), new Vector2(this.xPositionOnScreen - 4 + IClickableMenu.spaceToClearSideBorder + this.width / 3 + this.width / 3, this.yPositionOnScreen + IClickableMenu.spaceToClearTopBorder + 128 + 8 + 64 * (index + 1)), this.ThirdProfessionColor);
            }

            this.drawMouse(b);
        }
    }
}


testing.cs
//Global Variables
public readonly Vector2[] bigKegsInput = new Vector2[] { new Vector2(20,3), new Vector2(23,3), new Vector2(26,3), new Vector2(29,3), new Vector2(32,3) };
public readonly Vector2[] bigKegsOutput = new Vector2[] { new Vector2(20,6), new Vector2(23,6), new Vector2(26,6), new Vector2(29,6), new Vector2(32,6) };


//Inside Entry Method
InputEvents.ButtonPressed += InputEvents_ButtonPressed;
MenuEvents.MenuClosed += MenuEvents_MenuClosed;
TimeEvents.TimeOfDayChanged += TimeEvents_TimeOfDayChanged;

//Other Methods Inside Class
public bool IsKegable(Item _item)
{
    int _index = _item.ParentSheetIndex;
    int _category = _item.Category;
    return ( _index == 262 || _index == 304 || _index == 340 || _index == 433 || _category == -79 || _category == -75 );
}

public bool IsBigKegInput(Vector2 _position)
{
    return this.bigKegsInput.Contains(_position);
}

public bool IsBigKegOutput(Vector2 _position)
{
    return this.bigKegsOutput.Contains(_position);
}

public void SetKegAnimation(Layer _layer, Vector2 _tileLocation, TileSheet _tilesheet, int[] _tileIDs, long _interval)
{
    _layer.Tiles[(int)_tileLocation.X, (int)_tileLocation.Y] = new AnimatedTile(_layer, MakeAnimatedTile(_layer, _tilesheet, _tileIDs), _interval);
}

public StaticTile[] MakeAnimatedTile(Layer _layer, TileSheet _tilesheet, int[] _tileIDs)
{
    StaticTile[] _output = new StaticTile[_tileIDs.Count()];
    for(int i = 0; i < _tileIDs.Count(); i++)
    {
        _output[i] = new StaticTile(_layer, _tilesheet, BlendMode.Alpha, _tileIDs[i]);
    }
    return _output;
}

private void InputEvents_ButtonPressed(object sender, EventArgsInput e)
{
    if( Game1.currentLocation != null && Game1.currentLocation.mapPath.Value == "Maps\\Winery" && IsBigKegInput(e.Cursor.GrabTile) )
    {
        if( e.IsActionButton )
        {
            GameLocation _winery = Game1.currentLocation;                    
            Vector2 _chestLocation = new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y - 34);
            Vector2 _outputChestLocation = new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y - 24);

            Layer _layerBuildings = _winery.map.GetLayer("Buildings");
            Layer _layerFront = _winery.map.GetLayer("Front");
            TileSheet _tilesheet = _winery.map.GetTileSheet("bucket_anim");


            if ( !_winery.Objects.ContainsKey( _chestLocation ) )
            {
                Chest _newChest = new Chest(true){TileLocation = _chestLocation, Name = "|ignore| input chest for big ol kego"};
                _winery.Objects.Add(_chestLocation, _newChest);
            }
            if ( !_winery.Objects.ContainsKey(_outputChestLocation) )
            {
                Chest _newChest = new Chest(true) { TileLocation = _outputChestLocation, Name = "|ignore| output chest for big ol kego" };
                _winery.Objects.Add(_outputChestLocation, _newChest);
            }
            if( Game1.player.CurrentItem != null && IsKegable(Game1.player.CurrentItem) )
            {
                Chest _inputChest = (Chest)_winery.Objects[_chestLocation];
                Chest _outputChest = (Chest)_winery.Objects[_outputChestLocation];
                Item _item = Game1.player.CurrentItem;
                Item _remainder = null;
                switch (_item.ParentSheetIndex)
                {
                    case 262:
                        _item = new SObject(Vector2.Zero, 346, "Beer", false, true, false, false) { Name = "Beer" };
                        ((SObject)_item).setHealth(1750);
                        _remainder = _inputChest.addItem(_item);
                        SetKegAnimation(_layerBuildings, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 2), _tilesheet, new int[] { 4, 5, 6 }, 250);
                        SetKegAnimation(_layerFront, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 2), _tilesheet, new int[] { 18, 19, 20 }, 250);
                        if (_outputChest.items.Count <= 0)
                        {
                            SetKegAnimation(_layerBuildings, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 3), _tilesheet, new int[] { 11, 12, 13 }, 250);
                        }
                        break;
                    case 304:
                        _item = new SObject(Vector2.Zero, 303, "Pale Ale", false, true, false, false) { Name = "Pale Ale" };
                        ((SObject)_item).setHealth(2250);
                        _remainder = _inputChest.addItem(_item);
                        SetKegAnimation(_layerBuildings, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 2), _tilesheet, new int[] { 4, 5, 6 }, 250);
                        SetKegAnimation(_layerFront, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 2), _tilesheet, new int[] { 18, 19, 20 }, 250);
                        if (_outputChest.items.Count <= 0)
                        {
                            SetKegAnimation(_layerBuildings, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 3), _tilesheet, new int[] { 11, 12, 13 }, 250);
                        }
                        break;
                    case 340:
                        _item = new SObject(Vector2.Zero, 459, "Mead", false, true, false, false) { Name = "Mead" };
                        ((SObject)_item).setHealth(600);
                        _remainder = _inputChest.addItem(_item);
                        SetKegAnimation(_layerBuildings, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 2), _tilesheet, new int[] { 4, 5, 6 }, 250);
                        SetKegAnimation(_layerFront, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 2), _tilesheet, new int[] { 18, 19, 20 }, 250);
                        if (_outputChest.items.Count <= 0)
                        {
                            SetKegAnimation(_layerBuildings, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 3), _tilesheet, new int[] { 11, 12, 13 }, 250);
                        }
                        break;
                    case 433:
                        _item = new SObject(Vector2.Zero, 395, "Coffee", false, true, false, false) { Name = "Coffee" };
                        _item.Stack = (_item.Stack / 5) - (_item.Stack % 5);
                        ((SObject)_item).setHealth(120);
                        _remainder = _inputChest.addItem(_item);
                        SetKegAnimation(_layerBuildings, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 2), _tilesheet, new int[] { 1, 2, 3 }, 250);
                        SetKegAnimation(_layerFront, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 2), _tilesheet, new int[] { 15, 16, 17 }, 250);
                        if (_outputChest.items.Count <= 0)
                        {
                            SetKegAnimation(_layerBuildings, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 3), _tilesheet, new int[] { 8, 9, 10 }, 250);
                        }
                        break;
                    default:
                        switch (_item.Category)
                        {
                            case -79:
                                _item = new SObject(Vector2.Zero, 348, _item.Name + " Wine", false, true, false, false) { Name = _item.Name + " Wine" };
                                ((SObject)_item).Price = ((SObject)_item).Price * 3; 
                                helper.Reflection.GetField<SObject.PreserveType>(_item, "preserve").SetValue( SObject.PreserveType.Wine );
                                helper.Reflection.GetField<int>(_item, "preservedParentSheetIndex").SetValue(_item.ParentSheetIndex );
                                ((SObject)_item).setHealth(10000);
                                _remainder = _inputChest.addItem(_item);
                                SetKegAnimation(_layerBuildings, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 2), _tilesheet, new int[] { 1, 2, 3 }, 250);
                                SetKegAnimation(_layerFront, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 2), _tilesheet, new int[] { 15, 16, 17 }, 250);
                                if (_outputChest.items.Count <= 0)
                                {
                                    SetKegAnimation(_layerBuildings, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 3), _tilesheet, new int[] { 8, 9, 10 }, 250);
                                }
                                break;
                            case -75:
                                _item = new SObject(Vector2.Zero, 350, _item.Name + " Juice", false, true, false, false) { Name = _item.Name + " Juice" };
                                ((SObject)_item).Price = (int)((double)((SObject)_item).Price * 2.5);
                                helper.Reflection.GetField<SObject.PreserveType>(_item, "preserve").SetValue(SObject.PreserveType.Juice);
                                helper.Reflection.GetField<int>(_item, "preservedParentSheetIndex").SetValue(_item.ParentSheetIndex);
                                ((SObject)_item).setHealth(6000);
                                _remainder = _inputChest.addItem(_item);
                                SetKegAnimation(_layerBuildings, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 2), _tilesheet, new int[] { 1, 2, 3 }, 250);
                                SetKegAnimation(_layerFront, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 2), _tilesheet, new int[] { 15, 16, 17 }, 250);
                                if (_outputChest.items.Count <= 0)
                                {
                                    SetKegAnimation(_layerBuildings, new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y + 3), _tilesheet, new int[] { 8, 9, 10 }, 250);
                                }
                                break;
                        }
                        break;
                }
                if (_remainder == null)
                {
                    Game1.player.removeItemFromInventory(_item);
                }
            }                    
        }                
    }
    else if ( Game1.currentLocation != null && Game1.currentLocation.mapPath.Value == "Maps\\Winery" && IsBigKegOutput(e.Cursor.GrabTile) )
    {
        if( e.IsActionButton)
        {
            GameLocation _winery = Game1.currentLocation;
            Vector2 _chestLocation = new Vector2(e.Cursor.GrabTile.X, e.Cursor.GrabTile.Y - 27);
            Chest _chest = (Chest)_winery.Objects[_chestLocation];
            Game1.activeClickableMenu = new ItemGrabMenu(
               inventory: _chest.items,
               reverseGrab: false,
               showReceivingMenu: true,
               highlightFunction: InventoryMenu.highlightAllItems,
               behaviorOnItemSelectFunction: _chest.grabItemFromInventory,
               message: null,
               behaviorOnItemGrab: _chest.grabItemFromChest,
               canBeExitedWithKey: true, showOrganizeButton: true,
               source: ItemGrabMenu.source_chest,
               context: _chest
           );
        }               
    }
}

private void MenuEvents_MenuClosed(object sender, EventArgsClickableMenuClosed e)
{
    if(e.PriorMenu is ItemGrabMenu && Game1.currentLocation != null && Game1.currentLocation.mapPath.Value == "Maps\\Winery")
    {
        GameLocation _winery = Game1.currentLocation;
        Layer _layerBuildings = _winery.map.GetLayer("Buildings");
        TileSheet _tilesheet = _winery.map.GetTileSheet("bucket_anim");
        foreach(SObject o in _winery.Objects.Values)
        {
            if(o is Chest _chest && _chest.Name.Equals("|ignore| output chest for big ol kego") && _chest.items.Count <= 0)
            {
                _layerBuildings.Tiles[(int)_chest.TileLocation.X, (int)_chest.TileLocation.Y + 27] = new StaticTile(_layerBuildings, _tilesheet, BlendMode.Alpha, 21);
            }
        }
    }
}

public bool IsWineJuiceOrCoffee(Item _item)
{
    return (_item.ParentSheetIndex == 395 || _item.ParentSheetIndex == 348 || _item.ParentSheetIndex == 350);
}

public bool IsBeerBeerOrBeer(Item _item)
{
    return (_item.ParentSheetIndex == 346 || _item.ParentSheetIndex == 303 || _item.ParentSheetIndex == 459);
}

public void TimeEvents_TimeOfDayChanged(object sender, EventArgsIntChanged e)
{
    foreach (Building b in Game1.getFarm().buildings)
    {
        if (b.indoors.Value != null && b.buildingType.Value.Equals("Winery")) {
            GameLocation _winery = b.indoors.Value;
            Layer _layerBuildings = _winery.map.GetLayer("Buildings");
            Layer _layerFront = _winery.map.GetLayer("Front");
            TileSheet _tilesheet = _winery.map.GetTileSheet("bucket_anim");
            foreach (SObject o in b.indoors.Value.Objects.Values)
            {
                if (o is Chest _chest && _chest.Name.Equals("|ignore| input chest for big ol kego"))
                {
                    foreach (SObject _item in _chest.items)
                    {
                        if (_item.getHealth() > 0)
                        {
                            _item.setHealth(_item.getHealth() - 10);
                        }
                        else if (_item.getHealth() <= 0)
                        {
                            Chest _outputChest = ((Chest)b.indoors.Value.Objects[new Vector2(_chest.TileLocation.X, _chest.TileLocation.Y + 10)]);
                            Item _remainder = _outputChest.addItem(_item);
                            if (_remainder == null)
                            {
                                _chest.items.Remove(_item);
                            }
                            if (IsWineJuiceOrCoffee(_item))
                            {
                                SetKegAnimation(_layerBuildings, new Vector2(_chest.TileLocation.X, _chest.TileLocation.Y + 37), _tilesheet, new int[] { 22, 23, 24 }, 250);
                            }
                            else
                            {
                                SetKegAnimation(_layerBuildings, new Vector2(_chest.TileLocation.X, _chest.TileLocation.Y + 37), _tilesheet, new int[] { 25, 26, 27 }, 250);
                            }
                        }
                    }
                    if (_chest.items.Count <= 0)
                    {
                        _layerBuildings.Tiles[(int)_chest.TileLocation.X, (int)_chest.TileLocation.Y + 36] = new StaticTile(_layerBuildings, _tilesheet, BlendMode.Alpha, 0);
                        _layerFront.Tiles[(int)_chest.TileLocation.X, (int)_chest.TileLocation.Y + 36] = new StaticTile(_layerFront, _tilesheet, BlendMode.Alpha, 14);
                    }
                }
            }
        }
    }
}
